### PR TITLE
Fix for Internal Error: refDTypep points to dead module -- typedef through a parameterized sub-interface (regression) (#8124)

### LINKED ISSUE #8124: Internal Error: refDTypep points to dead module -- typedef through a parameterized sub-interface (regression)
While chasing something else I went back and ran a handful of small local tests left over from my earlier DepGraph work - that I haven't had time to get back to ><.   These are from  a suite of test-cases that I was using to probe the DepGraph (different ordering, execution, boundary conditions, etc) and I kept them out of the Verilator repo to avoid cluttering it with a large pile of near-duplicate small cases (which annoys folks ^^).

Anyway they all were passing against master at one time and I had thought that I had the same coverage in the existing CI tests that were part of the ledger work.  Unfortunately I was incorrect.  Several of them no longer pass. They were passing before #7943 (which yes I approved) I believe from initial investigation.

Here is an example of one of the test-cases that fails on latest master:

```
package a_pkg;
   typedef struct packed {
      int unsigned p_a;
   } cfg_t;
endpackage

interface sub_types #(parameter a_pkg::cfg_t cfg = 0);
   typedef logic [cfg.p_a-1:0] data_t;
endinterface

interface sub_if #(parameter a_pkg::cfg_t cfg = 0);
   sub_types #(cfg) sub_types();
   typedef sub_types.data_t data_t;
   typedef struct packed {
      logic [3:0] addr;
      data_t data;
   } data2_t;
endinterface

module sub (sub_if io);
endmodule

module t();
   parameter a_pkg::cfg_t cfg = '{p_a: 16};

   sub_if #(cfg) sub_io();

   sub u_sub(.io(sub_io));

   typedef sub_io.data2_t data2_t;
   localparam int COUNT = $bits(data2_t);

   initial begin
      if (COUNT != 20) $stop;
      $write("*-* All Finished *-*\n");
      $finish;
   end
endmodule
```

```
verilator --lint-only on current master (v5.050-192-gd2f62fda9):

%Error: Internal Error: t.v:19:22: ../V3LinkDotIfaceCapture.cpp:308: refDTypep of ''data_t'' points to dead module 'sub_types'
   19 |    typedef sub_types.data_t data_t;
      |                      ^~~~~~

```
Expected output, elaborates cleanly; $bits(data2_t) is 20. That is what happens on:

- 5.049 v5.048-24-g4a35cd28a -- clean
- 5.047 v5.046-135-g086bf351f -- clean

I will go do some archaeology here, and @solomatnikov its possible that this would also cover what you are seeing in your latest issue.  I don't know yet.

This will likely take me a couple days, but it also hopefully lights a fire for me to resume looking at my parallel DepGraph which was designed to eliminate all of the different bridges, corner cases, etc, that continually are popping up.

Thanks, -e-

### PR DESCRIPTION
**Pull request by em2machine**
Fixes #8124

### CODE REVIEW

File: src/V3Width.cpp [hunk 1]
```diff
@@ -2237,6 +2237,21 @@ class WidthVisitor final : public VNVisitor {
         case VAttrType::DIM_SIZE: {
             AstNodeDType* const dtypep = fromDTypep(nodep->fromp());
             UASSERT_OBJ(dtypep, nodep, "Unsized expression");
+            // A module that is still being copied does not have its final sizes.
+            {
+                const AstNodeModule* ownModp = nullptr;
+                for (AstNode* curp = dtypep; curp; curp = curp->aboveLoopp()) {
+                    if ((ownModp = VN_CAST(curp, NodeModule))) break;
+                }
```
Comment (wsnyder) [inline src/V3Width.cpp]: This is O(n^2) with tree size, can it be avoided or at least memoized?

**Comment by em2machine** [comment]
@wsnyder 

Agreed, I went around a bit on this one.  Tried to also use abovep() but this asserts unless at head or tail.  aboveLoopp() scans back all the statements and then walks upwards I thought?  I can add a guard if (m_paramsOnly) {}   This should remove the walk from that pass over the whole design.

I did try another direction where I added a bit essentially to every dtype in a template at clone time that I could use to disambiguate, but this also caused other issues.

A question for you: I have five or six more of these testcases to get through before I can get back on to master locally.   None of them are five minute debugs - i.e. this is the first one that I've gotten through.  I could use this PR to aggregate fixes for all of them (maybe a better approach comes along through that debug), or I can keep doing one at a time - but might have to rip up, etc.  I don't have a good sense yet. Opinions?

File: src/V3Width.cpp (hunk 1 above)
Comment (em2machine) [reply src/V3Width.cpp]: I can add a guard if (m_paramsOnly) {} This should remove the walk from that pass over the whole design.  See my comment below.  

```
            if (m_paramsOnly) {
                const AstNodeModule* ownModp = nullptr;
                for (AstNode* curp = dtypep; curp; curp = curp->aboveLoopp()) {
                    if ((ownModp = VN_CAST(curp, NodeModule))) break;
                }
                if (ownModp && ownModp->parameterizedTemplate() && !ownModp->dead()) {
                    nodep->dtypeSetInt();
                    return;
                }
            }
```

Comment (wsnyder) [reply src/V3Width.cpp]: Much better, but I think can still memoize the ownModp result for a given node, then it will be back to O(1) in most cases.  Actually I recall we might have already done such a back search otherwise elsewhere, see if there can be a common function for this....

**Comment by wsnyder** [comment]
If you think that the progress on these can be merged separately that's generally easier to review.  But if all of them are touching the exact same spots and the resulting PR src/ changes won't be much larger you can combine.

File: src/V3Width.cpp (hunk 1 above)
Comment (em2machine) [reply src/V3Width.cpp]: Yes, I could build the following, roughly mirroring containingClass():

```
std::unordered_map<const AstNode*, const AstNodeModule*> m_containingModp;

const AstNodeModule* containingModule(AstNode* nodep) {
}
```

So then it collapses to `const AstNodeModule* ownModp = containingModule(dtypep);`


I did look to find an existing walk that I could reuse / augment (and there are 9 or 10 instances of different walks ><) - and this in V3Param is the closest IMO.  I put on my mental list a todo to see if any of these can be aggregated without significant breakage.

Find which module a variable belongs to, using cache to avoid repeated backp() walks:
`AstNodeModule* findVarModule(AstVar* varp) {}`

Pretty close but:

1. Wrong input type. 
2. Wrong traversal for our input.
3. Wrong point in time (the cached output). 

I’ll prototype containingModule()

Comment (em2machine) [reply src/V3Width.cpp]: Ok, there are a couple of options.  

a) Build at the module level.  This ends up with machinery in processWorkQ()  (I'm somewhat sensitive to this as there are other ordering things, etc, there)
b) Build at the statement level.

a) Builds per module - which makes sense as this happens during parameterization and you'd like to get additional reuse of say $bits(data_t) if it occurs in multiple places.  Initial placement in the map, saves a walk every time the DIM is encountered.
b) Builds per statement - doesn't touch anything at the module level (reduction in error surface area), saves 1/2 the walks.  

So yes we can build the cache (and it does save usually half the walks).  I built both a) and b) with instrumentation in order to look and see how often module versus statement is useful.  There is value in module.

In the simple functional test-case here two walks versus 16.    I.e. cross statement re-use.  Of course if you have only a single statement then there is no re-use but you did save 1/2 the walks.

```
Mode        Queries     Walks     Levels     'Hops'
No cache    16          16        32         18
Statement   16          8         16         9
Module      16          2         4          3
```

I'm going to commit the module level for you to look at.

Comment (em2machine) [reply src/V3Width.cpp]: On a fairly large design.   Pretty much shows that statement level caching doesn't buy much on a non-trivial design.  Saves a few hundred out of almost 10 million hops...  So on the trivial test-cases at the statement level sure it saves 1/2 the walks but at least on this SOC it doesn't buy anything at that level.

image

I'm going to keep looking for additional opportunities to cut it down...like not having to go through the entire type table list...

Comment (em2machine) [reply src/V3Width.cpp]: Another simple optimization IMO.  Since I'm concerned with searching to find the parent I was using the 'walk the list backwards and then upwards', but the reality is that where I am starting in the list its much more efficient to 'walk forwards and then upwards'.  This results in the below.

image

In the latest commit.  There are probably other places (just a quick search shows several) where this might be useful in terms of walking - if one is not searching for statements that are previous but looking for the parent...

**Comment by em2machine** [comment]
@wsnyder I've decided to go through one by one in terms of my local DepGraph fails.  This should be ready for you.  I've merged master and should be current.  Thanks.

File: src/V3LinkDotIfaceCapture.cpp [hunk 2]
```diff
@@ -340,6 +360,27 @@ AstNodeModule* V3LinkDotIfaceCapture::findOwnerModule(AstNode* nodep) {
...
+std::unordered_map<const AstNode*, const AstNodeModule*> V3LinkDotIfaceCapture::s_containingModp;
+
+void V3LinkDotIfaceCapture::clearContainingModuleCache() { s_containingModp.clear(); }
+
+// The end of a list knows the parent, and the types we ask about are added at the end.
+static AstNode* parentFromTail(AstNode* nodep) {
+    AstNode* tailp = nodep;
+    while (tailp->nextp()) tailp = tailp->nextp();
+    return tailp->abovep();
+}
```
Comment (wsnyder) [inline src/V3LinkDotIfaceCapture.cpp:372]: There should be an Ast:: fuction that is similar.  If not, that suggests there's a better way.

File: src/V3LinkDotIfaceCapture.h [hunk 3]
```diff
@@ -128,13 +128,21 @@ class V3LinkDotIfaceCapture final {
 
     static CapturedMap s_map;
     static bool s_enabled;
+    // Which module each node sits in, only good while the tree holds still
+    static std::unordered_map<const AstNode*, const AstNodeModule*> s_containingModp;
```
Comment (wsnyder) [inline src/V3LinkDotIfaceCapture.h:132]: Should V3Broken check this cache for stale entries, or otherwise have asserts to make sure some future change doesn't break what's being kept here?  Generally we avoid static state.

For that matter maybe this is better held in AstNetlist so is obvious global netlist information.

**Review by wsnyder — CHANGES_REQUESTED** [review]
See changes above.  Please also rebase, and re-request review when you think ready, thanks

File: src/V3AstNodes.cpp [hunk 4]
```diff
@@ -1950,7 +1950,22 @@ void AstNetlist::astConstOrigParamName(const AstConst* nodep, const string& name
 void AstNetlist::astConstOrigParamNameErase(const AstConst* nodep) {
     m_constOrigParamNames.erase(nodep);
 }
+const AstNodeModule* AstNetlist::containingModule(const AstNode* nodep) {
+    if (const AstNodeModule* const modp = VN_CAST(nodep, NodeModule)) return modp;
+    const auto it = m_containingModules.find(nodep);
+    if (it != m_containingModules.end()) return it->second;
+    // Only true parents are followed.
+    AstNode* const abovep = nodep->aboveTailp();
```
Comment (wsnyder) [inline src/V3AstNodes.cpp:1958]: On what basis do you believe that this node is likely to be closer to the tail than to the head?

**Review by wsnyder — CHANGES_REQUESTED** [review]

File: src/V3AstNodes.cpp (hunk 4 above)
Comment (em2machine) [reply src/V3AstNodes.cpp:1958]: Measurement on very large designs (complete SOCs, not IPs).  Heres an example Instrumentation:
```
- aboveLoopp(): 8,473,461 peer steps (~7,394/walk)
- aboveTailp(): 476,959 peer steps (~416/walk)
```
Can I give a hard guarantee that its always cheaper to walk to the tail versus the head?  Nope, but in everything I've measured there is a heavy bias towards these being towards the tail.

This is a performance optimization based on the profiling work I was doing here before I took a break.  Using aboveLoopp() is functionally equivalent but lower perf in everything I've tested, so I can replace with aboveLoopp() if you prefer and delete aboveTailp().  At the full Verilation level this is a very small contributor to the CPI stack, so its a micro-optimization...

Comment (wsnyder) [reply src/V3AstNodes.cpp:1958]: Great data!  Can you run an experiment to always use the aboveTailp algorithm instead of aboveLoopp?

Comment (em2machine) [reply src/V3AstNodes.cpp:1958]: I instrumented aboveLoopp() to count both the backward distance to the head and the forward distance to the tail for every call.  

Interestingly enough on the large SOC that I was using as a test case there are no aboveLoopp() calls (it doesn’t have ‘disable’ and no classes).   I did verify that replacing aboveLoopp() with aboveTailp() is semantically equivalent - not suggesting this as a course of action, but wanted to make sure what I thought would be true was ^^.  

Here is some data from the test suite (which would be biased due to the nature of the tests):
```
- disable tests: 3464 calls, 7,733 back vs 16,235 fwd,  tail is 2.1x worse
- class tests: 1604 calls, 3,653 back vs 2,765 fwd,  tail is 0.76x, ~24% better
```

Comment (wsnyder) [reply src/V3AstNodes.cpp:1958]: Can you please submit a draft pr with just that and we'll run rtlmeter?

Comment (em2machine) [reply src/V3AstNodes.cpp]: So per the experiment, basically even.  I'll just switch to aboveLoopp (if we are agreed), rebase, and re-push.  Would like to get this one off my plate if possible as I have another internal crash that I'm looking at (another UAF I believe) that is likely dependent on this one and the previous UAF that I fixed a while ago.

Comment (em2machine) [reply src/V3AstNodes.cpp]: ok, so fix ends up being the cache (m_containingModules, cleared per module + at end of V3Param::param).  Walk direction was a nit at full-verilation (2nd or 3rd order effect).

File: src/V3LinkDotIfaceCapture.cpp (hunk 2 above)
Comment (em2machine) [reply src/V3LinkDotIfaceCapture.cpp]: Per our experiment, using existing method.

File: src/V3LinkDotIfaceCapture.h (hunk 3 above)
Comment (em2machine) [reply src/V3LinkDotIfaceCapture.h]: Now no static state the cache is now a AstNetlist member.

**Comment by em2machine** [comment]
@wsnyder I think this is ready to go now.  Looks like the CI sequence has changed so not 100% sure what is blocking the 'all-green'....

**Comment by gezalore** [comment]
Please merge/rebase to get it green, needs master changes to CI scripts. Also needs to have `pr: regression` for test to run, which I have added for you.

**Comment by em2machine** [comment]
Feeling dense.  Updated my fork, git fetch origin, did a git merge origin/master into my branch, ran locally to make sure all is good before I pushed.  What am I missing?
